074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+
+## Current situation
+
+The operator handles many configurations and operational aspects but lacks support for specifying additional volumes in the Custom Resource Definition (CRD). This limitation can restrict users from accessing log files or integrating with external storage solutions directly.
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.
+- Validation: Ensure that the additional volumes are validated correctly and do not conflict with existing volume definitions such as `/var/lib/kafka/` and `/tmp`.
+- Documentation: Update the Strimzi documentation to include examples and guidelines on how to use the new `additionalVolumes` field.
+- Security: Ensure that the additional volumes do not introduce security vulnerabilities by validating the configurations.
+- Testing: Comprehensive testing to cover various scenarios and use cases, ensuring that the feature works as expected.
+
+### Possible Delivery Mechanisms
+
+- CRD Updates: Modify the existing CRDs to include the new field.
+  - This will be done by extending the `PodTemplate.java` class to include the new `additionalVolumes` field as a list.
+- Operator Logic: Update the operator logic to handle the additional volumes during the deployment process.
+  - This should be possible to implement in the `KafkaCluster.java` class.
+
+The configuration could look as follows:
+
+```yaml
+    additionalVolumes:
+    - name: example-configmap
+      path: /path/to/mount/volume
+      #subPath: mykey # Add this to mount only a specific key of the configmap/secret
+      configMap:
+        name: config-map-name
+    - name: temp
+      path: /tmp/logs
+      emptyDir: {}
+    - name: example-csi-volume
+      path: /path/to/mount/volume
+      #subPath: "subpath" # Add this to mount the CSI volume at a specific subpath
+      csi:
+        driver: csi-driver-name
+        readOnly: true
+        volumeAttributes:
+          secretProviderClass: example-secret-provider-class
+    - name: example-projected-volume
+      path: /path/to/mount/volume
+      projected:
+        sources:
+          serviceAccountToken
+```
+
+### Supported volumes
+
+The following types of volumes are proposed as initial support. More types could be added later.
+
+- Secret
+- ConfigMap
+- EmptyDir
+- CSI
+- Projected volume

```

Comment:
It would be good to add some considerations about the security implications this has. Mounting Secrets or projected volumes is possibly dangerous as it allows you to mount things you otherwise might not have access to and use them. I'm not saying this is a blocker, but:
* It should be raised here for consideration while reviewing the proposal
* It should be documentad in the docs as part of the implementation if this is approved

Reply:
I am not sure why this is a concern from the operator point of view? It would be the ones deploying and operating kafka which has that responsbility and the RBAC setup in kubernetes.

Reply:
I agree that this isn't an RBAC issue, since Kubernetes RBAC can't be used to disallow a Pod from mounting a Secret. Regardless, I don't really see an issue here. It is a design choice by the Kubernetes developers that Pods are able to mount Secrets in the same namespace, so Kubernetes users should never place Secrets in a namespace where a Pod shouldn't have access to it.
Adding a note about a possible security issue also seems weird, since it is already possible to access arbitrary secrets in the namespace using several mechanisms in the CRD. For example, it is possible to use something like:

```yaml
template:
  kafkaContainer:
    env:
    - name: VERY_SECRET_SECRET
      valueFrom:
        secretKeyRef:
          name: my-secret
          key: some-key
    envFrom:
    - secretRef:
        name: my-secret
```
 
The `ContainterTemplate` documentation does not contain a security notice either: https://strimzi.io/docs/operators/latest/configuring#type-ContainerTemplate-reference

Reply:
I agree that this isn't an RBAC issue, since Kubernetes RBAC can't be used to disallow a Pod from mounting a Secret.
Regardless, I don't really see an issue here. It is a design choice by the Kubernetes developers that Pods are able to mount Secrets in the same namespace, so cluster admins should never place Secrets in a namespace where a Pod shouldn't have access to it.
Adding a note about a possible security issue also seems weird, since it is already possible to access arbitrary secrets in the namespace using several mechanisms in the different CRDs, and none of these ways have a warning about it

Reply:
I think this still should be covered here and in the future docs. As I said, I do not think it is a blocker and it obviously cannot be easily avoided unless we significantly limit what can be mounted. But it is a notable thing that should be documented in the docs in the future as well as here in the proposal.

> Adding a note about a possible security issue also seems weird, since it is already possible to access arbitrary secrets in the namespace using several mechanisms in the different CRDs, and none of these ways have a warning about it

Connect supports mounting custom Secrets or Config Maps through the CRD. Secrets are allowed also for TLS purposes. We do not support things such as Downward API. We also do not support it for all kind of containers. It is also less obvious and more hidden so it might not be clear on a quick look during a security review. So I think this is different.

TBH, I'm not sure I understand why it seems to be such a problem - I'm not blocking anything, I'm not asking to rewrite Kubernetes to prohibit it ... just asking for it to be properly documented in a paragraph or two.

Reply:
We will rework this. We will remove projected volumes and look more into CSI volumes vs. PVC.
As discussed in the community call, this should remove most security concerns.